[FEATURE]  Infrastructure for JavaScript modules and importmaps (#2455)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ApiOverview/Backend/JavaScript/ES6/Index.rst

ES6 in the TYPO3 Backend
========================

..  versionchanged:: 12.0
    Starting with TYPO3 v12.0 JavaScript ES6 modules may be used instead of

>     AMD modules, both in backend and frontend context.

Reviewer: Why mentioning frontend? I can use it also in v11 in frontend if I want. It sounds, that only v12 is capable of using ES6 in frontend - which is not true.

Author: It is written like that in the changelog entry

> JavaScript node-js style path resolutions are managed by

Reviewer: ```suggestion
JavaScript Node.js style path resolutions are managed by
```

"Node.js", see: https://nodejs.org/en/

But why mentioning Node.js? It is a standard which Node.js (and others) are implementing.

Author: It is written like that in the changelog entry, how would you call that

Reviewer: I don't understand what that part means (but I am not a JS guy) ... then leave it.

> `importmaps <https://wicg.github.io/import-maps/>`_, which allow web

Reviewer: The URL is a redirect to https://html.spec.whatwg.org/multipage/webappapis.html#import-maps

> `importmaps <https://wicg.github.io/import-maps/>`_, which allow web

Author: ```suggestion
`importmaps <https://html.spec.whatwg.org/multipage/webappapis.html#import-maps>`_, which allow web
```

pages to control the behavior of JavaScript imports.

> By the time of writing importmaps are supported natively by Google Chrome,

Reviewer: ```suggestion
By the time of writing, importmaps are supported natively by Google Chrome,
```

"By the time of writing": either add a date, like "November 2022" or add the browser versions you are referring to. The reader has no clue when it was written.

> In november 2022 importmaps are supported natively by Google Chrome,

Reviewer: ```suggestion
In November 2022 importmaps are supported natively by Google Chrome,
```

> By the time of writing importmaps are supported natively by Google Chrome,

Author: ```suggestion
In november 2022 importmaps are supported natively by Google Chrome,
```

> a polyfill is available for Firefox and Safari and included by TYPO3 core

Reviewer: ```suggestion
a polyfill is available for Firefox and Safari and included by TYPO3 Core
```

and applied whenever an importmap is emitted.

> For security reasons importmap configuration is only emitted when the modules

Reviewer: ```suggestion
For security reasons, importmap configuration is only emitted when the modules
```

are actually used, that means when a module has been added to the current

> page response via via :php:`PageRenderer->loadJavaScriptModule()` or

Reviewer: ```suggestion
page response via :php:`PageRenderer->loadJavaScriptModule()` or
```

:php:`JavaScriptRenderer->addJavaScriptModuleInstruction()`.
Exposing all module configurations is possible via
:php:`JavaScriptRenderer->includeAllImports()`, but that should only be

> done in backend context for logged in users, to avoid disclosing installed

Reviewer: ```suggestion
done in backend context for logged-in users to avoid disclosing installed
```

extensions to anonymous visitors.

Configuration
=============

A simple configuration example for an extension that maps
the `Public/JavaScript` folder to an import prefix `@vendor/my-extensions`:

..  literalinclude:: _JavaScriptModulesSimple.php
    :caption: EXT:my_extension/Configuration/JavaScriptModules.php

Complex configuration example containing recursive-lookup exclusions,
third-party library definitions and overwrites:

..  literalinclude:: _JavaScriptModulesExtended.php
    :caption: EXT:my_extension/Configuration/JavaScriptModules.php

.. _backend-javascript-es6-loading:

Loading ES6
===========

A module can be added to the current page response either via
:php:`PageRenderer` or as :php:`JavaScriptModuleInstruction` via
:php:`JavaScriptRenderer`:

..  literalinclude:: _PageRendererJavaScriptLoading.php
    :caption: EXT:my_extension/Classes/SomeNamespace/SomeClass.php

> In Fluid template the `includeJavaScriptModules` property of the

Reviewer: ```suggestion
In a Fluid template the `includeJavaScriptModules` property of the
```

:html:`<f:be.pageRenderer>` ViewHelper may be used:

..  literalinclude:: _BackendFluidTemplate.html
    :caption: EXT:my_extension/Resources/Private/Backend/Templates/SomeTemplate.html

Some tips on ES6
================

> In the TYPO3 Core there are no ES6 JavaScript files created directly. JavaScript

Reviewer: ```suggestion
No ES6 JavaScript files are created directly in the TYPO3 Core. JavaScript
```

is created as TypeScript module which is then converted to ES6 JavaScript

> during the build process. However TypeScript and ES6 are quite similar, you

Reviewer: ```suggestion
during the build process. However, TypeScript and ES6 are quite similar, you
```

can therefore look into those files for reference. The TypeScript files can be
found on GitHub at
`Build/Sources/TypeScript <https://github.com/TYPO3/typo3/tree/main/Build/Sources/TypeScript>`__.

> For an example on a ES6 JavaScript file have a look at the JavaScript example in

Reviewer: ```suggestion
For examples of an ES6 JavaScript file have a look at the JavaScript example in
```

Too much "example" in this sentence, maybe "have a look at the JavaScript part in LinkHandler Tutorial or in the Notification API".

Why "LinkHandler Tutorial" and not "link handler tutorial"?

the :ref:`LinkHandler Tutorial <tutorial_backend_link_handler_javascript>` or
the example in the :ref:`Notification API <notification_api>`.

> For a real life example on how to introduce ES6 modules into a large extension

Reviewer: ```suggestion
For a practical example on how to introduce ES6 modules into a large extension
```

"real life" sounds that other examples are just theoretical.

see this commit for EXT:news: `[TASK] Add support for TYPO3 v12 ES6
modules <https://github.com/bnf/news/commit/f8e196b67ceaa2f56699fbf464080dde668ad526>`__.

Using JQuery
------------

> There is a general tendency to remove dependencies to jQuery in the Core whenever

Reviewer: Hmm, would rephrase that, something like: "In the TYPO3 Core usage of jQuery is eliminated step-by-step as the functionality is provided by native JavaScript nowadays."

Author: Ok, go ahead

possible.

If you still have to use jQuery in your third-party extension, include it
with the following statement:

..  code-block:: javascript

    import $ from 'jquery';

.. _requirejs-migration:

Migration from RequireJS
========================

..  note:: For general information on how to migrate from AMD (RequireJS) to
    ES6 see this article: `Arthur Yidi: Migrate JavaScript Modules From AMD

>     to ES6 <https://arthuryidi.com/migrate-amd-modules/>`__. Here we talk about

Reviewer: "Here we talk ..." -> "This is about the TYPO3-specific details."

We don't talk here, we read ;-)

>     to ES6 <https://arthuryidi.com/migrate-amd-modules/>`__. Here we talk about

Author: ```suggestion
    to ES6 <https://arthuryidi.com/migrate-amd-modules/>`__. This is about 
```

    the TYPO3 specific details.

RequireJS is shimmed to prefer ES6 modules if available, allowing any extension
to ship ES6 modules by providing an importmap configuration in

> :file:`Configuration/JavaScriptModules.php` while providing full backwards

Reviewer: ```suggestion
:file:`Configuration/JavaScriptModules.php` while providing full backward
```

compatibility support for extensions that load modules via RequireJS.

Existing RequireJS modules can load new ES6 modules via a bridge that
prefers ES6 modules over traditional RequireJS AMD modules. This allows
extensions authors to migrate to ES6 without breaking dependencies that

> used to load a module of that extension via RequireJS.

Reviewer: ```suggestion
previously loaded a module of that extension via RequireJS.
```
